File: src/web_socket.py
import uuid
from fastapi import WebSocket, WebSocketDisconnect

# Local imports
from .pairing_manager import PairingManager


# Pairing manager instance
manager = PairingManager()

file_buffers = {}
active_connections = {}

# WebSocket endpoint
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    client_id = str(uuid.uuid4())
    active_connections[client_id] = websocket

    current_pairing_key = None

    try:
        while True:
            data = await websocket.receive_json()
            action = data.get("action")
            pairing_key = data.get("pairing_key")

            # === Pairing Logic ===
            if action == "register":
                manager.register(pairing_key, client_id)
                current_pairing_key = pairing_key
                await websocket.send_json({"type": "registered", "message": "Pairing key registered."})

            elif action == "connect":
                if manager.check_pair(pairing_key):
                    partner_id = manager.get_client(pairing_key)
                    partner_ws = active_connections.get(partner_id)
                    if partner_ws:
                        await partner_ws.send_json({
                            "type": "connected",
                            "partner_id": client_id
                        })
                        await websocket.send_json({
                            "type": "connected",
                            "partner_id": partner_id
                        })

                        # Track mutual connection
                        manager.pair_clients(client_id, partner_id)
                        current_pairing_key = pairing_key
                else:
                    await websocket.send_json({"type": "error", "message": "Invalid pairing key."})


            # === File Transfer Logic ===
            elif action == "send_file":
                partner_id = data.get("partner_id")
                filename = data.get("filename")
                filesize = data.get("filesize")

                partner_ws = active_connections.get(partner_id)
                if partner_ws:
                    await partner_ws.send_json({
                        "type": "file_offer",
                        "from": client_id,
                        "filename": filename,
                        "filesize": filesize
                    })

            elif action == "file_chunk":
                partner_id = data.get("partner_id")
                chunk = data.get("chunk")  # base64 encoded

                partner_ws = active_connections.get(partner_id)
                if partner_ws:
                    await partner_ws.send_json({
                        "type": "file_chunk",
                        "from": client_id,
                        "chunk": chunk
                    })

            elif action == "file_complete":
                partner_id = data.get("partner_id")
                partner_ws = active_connections.get(partner_id)
                if partner_ws:
                    await partner_ws.send_json({
                        "type": "file_complete",
                        "from": client_id
                    })

            elif action == "reject_file":
                sender_id = data.get("sender_id")
                sender_ws = active_connections.get(sender_id)
                if sender_ws:
                    await sender_ws.send_json({
                        "type": "file_rejected"
                    })

    except WebSocketDisconnect:
        active_connections.pop(client_id, None)

        partner_id = manager.get_partner(client_id)
        if partner_id:
            partner_ws = active_connections.get(partner_id)
            if partner_ws:
                try:
                    await partner_ws.send_json({
                        "type": "partner_disconnected",
                        "message": "Your partner has disconnected."
                    })
                except Exception as e:
                    print(f"Failed to notify partner: {e}")

        # Clean up all mapping
        manager.remove_pairing(client_id)

File: src/web_socket.js
const { v4: uuidv4 } = require('uuid');
const PairingManager = require('./pairing_manager');

// PairingManager instance
const manager = new PairingManager();

// Active WebSocket connections
const activeConnections = {};

// WebSocket handler function
function websocketHandler(ws) {
    const clientId = uuidv4();
    activeConnections[clientId] = ws;
    let currentPairingKey = null;

    ws.on('message', (msg) => {
        let data;
        try { data = JSON.parse(msg); } catch { return; }

        const { action, pairing_key, partner_id, filename, filesize, chunk, sender_id } = data;

        if (action === 'register') {
            manager.register(pairing_key, clientId);
            currentPairingKey = pairing_key;
            ws.send(JSON.stringify({ type: 'registered', message: 'Pairing key registered.' }));

        } else if (action === 'connect') {
            if (manager.checkPair(pairing_key)) {
                const partnerId = manager.getClient(pairing_key);
                const partnerWs = activeConnections[partnerId];
                if (partnerWs) {
                    partnerWs.send(JSON.stringify({ type: 'connected', partner_id: clientId }));
                    ws.send(JSON.stringify({ type: 'connected', partner_id: partnerId }));
                    manager.pairClients(clientId, partnerId);
                    currentPairingKey = pairing_key;
                }
            } else {
                ws.send(JSON.stringify({ type: 'error', message: 'Invalid pairing key.' }));
            }

        } else if (action === 'send_file') {
            const pw = activeConnections[partner_id];
            if (pw) {
                pw.send(JSON.stringify({
                    type: 'file_offer', from: clientId, filename, filesize
                }));
            }

        } else if (action === 'file_chunk') {
            const pw = activeConnections[partner_id];
            if (pw) {
                pw.send(JSON.stringify({
                    type: 'file_chunk', from: clientId, chunk
                }));
            }

        } else if (action === 'file_complete') {
            const pw = activeConnections[partner_id];
            if (pw) {
                pw.send(JSON.stringify({ type: 'file_complete', from: clientId }));
            }

        } else if (action === 'reject_file') {
            const senderWs = activeConnections[sender_id];
            if (senderWs) {
                senderWs.send(JSON.stringify({ type: 'file_rejected' }));
            }

        } else if (action === 'partner_disconnected') {
            const partner = manager.getPartner(clientId);
            if (partner) {
                const pw = activeConnections[partner];
                if (pw) {
                    pw.send(JSON.stringify({ type: 'partner_disconnected', message: 'Your partner has disconnected.' }));
                }
            }
            manager.removePairing(clientId);
            ws.close();
        }
    });

    ws.on('close', () => {
        delete activeConnections[clientId];
        const partner = manager.getPartner(clientId);
        if (partner) {
            const pw = activeConnections[partner];
            if (pw) {
                pw.send(JSON.stringify({ type: 'partner_disconnected', message: 'Your partner has disconnected.' }));
            }
        }
        manager.removePairing(clientId);
    });
}

module.exports = websocketHandler;
